NuMPI/Optimization/polonsky_keer.py

```diff
@@ -0,0 +1,232 @@
+
+    Parameters
+    ----------
+    objective : callable.
+                The objective function to be minimized.
+                            fun(x) -> float(energy),ndarray(gradient)
+                where x is the input ndarray.
+    hessp : callable
+            Function to evaluate the hessian product of the objective.
+            Hessp should accept either 1 argument (desscent direction) or
+            2 arguments (x,descent direction).
+                            hessp(des_dir)->ndarray
+                                    or
+                            hessp(x,des_dir)->ndarray
+            where x is the input ndarray and des_dir is the descent direction.
+
+    x0 : ndarray
+         Initial guess. Default value->None.
+         ValueError is raised if "None" is provided.
+
+    gtol : float, optional
+           Default value : 1e-8
+
+    mean_value : int/float, optional
+               If you want to apply the mean_value constraint then provide an
+               int/float value to the mean_value.
+
+    residual_plot : bool, optional
+                    Generates a plot between the residual and iterations.
+
+    maxiter : int,optional
+              Default, maxiter=5000
+              Maximum number of iterations after which the program will exit.
+
+    Returns
+    -------
+    success : bool
+              True if convergence else False.
+    x : array
+        array of minimized x.
+    jac : array
+          value of residual at convergence/non-convergence.
+    nit : int
+          Number of iterations
+    message : string
+              Convergence or Nodes_dir-Convergence
+
+    References
+    ----------
+    ..[1] I.A. Polonsky, L.M. Keer, Wear 231, 206 (1999)
+    """
+
+    if not isinstance(mean_value, (type(None), int, float)):
+        raise ValueError('Inappropriate type: {} for mean_value whereas a '
+                         'float \
+            or int is expected'.format(type(mean_value)))
+
+    if not isinstance(residual_plot, bool):
+        raise ValueError('Inappropriate type: {} for "residual_plot" whereas '
+                         'a bool \
+                         is expected'.format(type(residual_plot)))
+
+    gtol = gtol
+    fun = objective
+    x0 = x0.flatten()
+    if x0 is not None:
+        x = x0.copy()
+        # print("Total force is  :: {}".format(np.sum(x)))
+        delta = 0
+        G_old = 1
+    else:
+        raise ValueError('Input required for x0/initial value !!')
+
+    grads = np.array([])
+    iterations = np.array([])
+    rms_pen_ = np.array([])
+
+    des_dir = np.zeros(np.shape(x))
+
+    if residual_plot:
+        grads = np.append(grads, 0)
+        iterations = np.append(iterations, 0)
+
+    n_iterations = 1
+
+    while (n_iterations < maxiter + 1):
+
+        '''Mask to truncate the negative values'''
+        mask_neg = x <= 0
+        x[mask_neg] = 0.0
+
+        '''Initial residual or GAP for polonsky-keer'''
```

Comment:
It's ok to use contact mechanics terminology in the comments to make the parallel to the paper.